### PR TITLE
Upgraded OPA and added policy-per-record auth to getting a record

### LINKED ISSUE #2362: Add authorization to getting a record's top-level metadata (name/id)
### Problem description
Protect the [/records/{id}](https://dev.magda.io/api/v0/apidocs/#api-Registry_Record_Service-GetV0RegistryRecordsId) endpoint with authorization - don't worry about protecting the aspects (yet), this is simply about whether a call to get a certain record should return `200` or `404.

However to get this working requires a fair bit of fundamentals to be done - setting up a policy in OPA, and setting up the plumbing to make the registry know how to talk to OPA.

Record authentication wasn't done at all in [the PoC PR](https://github.com/magda-io/magda/pull/2325), so this diverges a bit from that, but you can use a lot of what's done for record aspects.

### Acceptance Criteria
- All records should have some kind of read policy - this should be either a policy that allows all users with a view record permission to view it, or one that only allows access from an owning org unit (see technical notes below)
- Existing records should be accessible just as they are now (so data.gov.au doesn't break)
- If a record can be read, it should return the name, id and any aspects requested (no need for auth on aspects yet)
- If a record cannot be read, it should return 404

### Technical Notes
- A new `access-control` aspect needs to be created that contains a `orgUnitOwnerId` string.
- New policies need to be written for OPA - one that allows any user with the `object/registry/record/read` to view, and one that requires them to both have that permission AND to have the `orgUnitOwnerId` in the `access-control` aspect match the user's `managingOrgUnitIds` (see [these](https://github.com/magda-io/magda/tree/issues/2247-registry-auth/deploy/helm/magda/charts/opa/policies/object/registry/aspect))
- A new column needs to be added to `records` in the registry db - `authpolicyread` - which will contain a reference to the OPA policy to be queried. Make sure that existing rows end up with a read value that allows them to be read just as they would have before.
- Functionality needs to be added to query OPA via the Auth API. For this use case you can either use partial evaluation like the PoC PR does (set the record's details to "unknown", ask OPA what you need to know about a record to allow/deny access and then turn that into SQL), or you could retrieve all a record's aspects and pass them directly into OPA, which will save having to implement an OPA parser. Don't forget that a policy can be set which uses aspects that aren't being requested (i.e. the `dcat-dataset-strings` aspect could have a policy that requires the `access-control` aspect to be read in order to allow/deny access), so you can't only pass the aspect that's being requested into OPA.

### Out of Scope
- Any aspect-specific authorisation (although that will come later) - just return everything for now whether its authorised or not
- Definitely any authorisation around dereferenced aspects
- Getting a single record's details via other endpoints (e.g. history, summary etc)
- Reading multiple record aspects or multiple records
- Anything to do with update/create/delete of anything
- Anything to do with search
- Any kind of admin for any of this
- Actually creating or registering the `access-control` aspect

This one small step, giant leaps can come later ;)

### PR DESCRIPTION
**Pull request by AlexGilleran**
### What this PR does

Fixes #2362 

- Makes auth work in the way described in #2362 
    - Every record now has an `authnReadPolicyId` column, that determines what policy to look up in OPA to authorise access to that record
    - The existing policy id set in config now determines a default fallback policy if this column is null (so that existing implementations, particularly the digital twin, still work).
    - The fallback policy, by default, is a new one - "public" - that just allows everything, a la data.gov.au etc
- Upgraded OPA to the latest version. On the upside, this new version falls back to "support" way less and uses queries way more, on the downside this broke the existing Content API authn, so I've rewritten the parser for that to rely on `queries` and not `support` like the registry api one does. If we had complex content policies this could be a problem but we don't so it's OK for now (I think, please try to prove me wrong!)
- The old views that were used for determining whether something is public are now gone because `authnReadPolicyId = public` now does that.

### Warts
- Because the content api tests mock out the `Database` class, and the OPA logic is all in there, essentially none of the OPA changes to the content api are tested. We should resolve this... but later, given that progressing registry auth is way more important than progressing content auth.

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: magda-registry-api/README.md [hunk 1]
```diff
@@ -51,7 +51,7 @@ users in different organization units. See the comment in the sql file for detai
 ### Start opa server
 
 ```
-cd deploy/helm/magda-core/charts/opa
+cd deploy/helm/magda/charts/magda-core/charts/opa
```
Comment (t83714) [inline magda-registry-api/README.md]: `magda-core` chart files are not in `magda` 's chart directory so original path should be correct?

Comment (AlexGilleran) [reply magda-registry-api/README.md]: Oops yeah... I moved them and moved them back, looks like I missed a spot. Good catch :)

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RegistryApp.scala [hunk 2]
```diff
@@ -60,9 +60,9 @@ object RegistryApp extends App {
   logger.info("Starting MAGDA Registry")
 
   GlobalSettings.loggingSQLAndTime = LoggingSQLAndTimeSettings(
```
Comment (t83714) [inline magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RegistryApp.scala]: Would it be easier if we set debug logging in [application.conf](https://github.com/magda-io/magda/blob/45d5f800449aee5cc676f0c921906cd2e2f43615/magda-registry-api/src/main/resources/application.conf) ?
e.g.:
```
scalikejdbc{
    global {
        loggingSQLAndTime.enabled = true
        loggingSQLAndTime.singleLineMode = true
        loggingSQLAndTime.logLevel = DEBUG
        loggingSQLAndTime.warningEnabled = true
        loggingSQLAndTime.warningThresholdMillis = 5
        loggingSQLAndTime.warningLogLevel = warn
    }
}
```
In this way, we probably can override the log setting through runtime parameter without re-compiling the registry-api / docker image.

File: magda-registry-api/README.md [hunk 3]
```diff
@@ -51,7 +51,7 @@ users in different organization units. See the comment in the sql file for detai
 ### Start opa server
 
 ```
-cd deploy/helm/magda-core/charts/opa
+cd deploy/helm/magda/magda-core/charts/opa
```
Comment (t83714) [inline magda-registry-api/README.md]: probably should be `deploy/helm/magda-core/charts/opa` (without /magda/).

File: magda-typescript-common/src/OpaCompileResponseParser.ts [hunk 4]
```diff
@@ -968,33 +970,68 @@ export default class OpaCompileResponseParser {
...
+            (!_.isArray(this.data.support) || !this.data.support.length)
+        ) {
             // --- mean no rule matched
             return [];
         }
+
+        const queries = this.data.queries;
+
+        if (queries) {
+            queries.forEach((q: any, i: number) => {
```
Comment (t83714) [inline magda-typescript-common/src/OpaCompileResponseParser.ts]: Probably easier (and more generic as we can handle any response no matter whether `support` / `queries` are there or not) to convert `this.data.queries` (answer) to a package.
Compared with packages in `support`, the answer/queries section contains rules body without package namespace.
We probably can put all answer queries into a fake namespace `magda.query.answer` (package path `magda.query` & fake rule name `answer`).
And add it to the support packages (if support is empty, we just create an array with package name `magda.query.answer`) so the same rule drying logic can be applied.
Then, we can call `evaluateRule` the fake rule `magda.query.answer`

Ignore me if the current code works well already 😄

File: deploy/helm/magda-core/charts/opa/templates/deployment.yaml [hunk 5]
```diff
@@ -17,7 +17,7 @@ spec:
 {{- end }}
       containers:
       - name: opa
-        image: openpolicyagent/opa:0.10.7
+        image: openpolicyagent/opa:0.17.1
```
Comment (t83714) [inline deploy/helm/magda-core/charts/opa/templates/deployment.yaml]: We probably need to update deploy/helm/magda-core/charts/opa/docker-compose.yml as well

Comment (AlexGilleran) [reply deploy/helm/magda-core/charts/opa/templates/deployment.yaml]: 👍

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RegistryApp.scala (hunk 2 above)
Comment (AlexGilleran) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RegistryApp.scala]: 😬 didn't mean to commit this but that sounds good. Although I think we can already override it, but having it in application.conf makes sense, I'll make it so 👍

File: magda-registry-api/README.md (hunk 3 above)
Comment (AlexGilleran) [reply magda-registry-api/README.md]: double oops 😬

File: magda-typescript-common/src/OpaCompileResponseParser.ts (hunk 4 above)
Comment (AlexGilleran) [reply magda-typescript-common/src/OpaCompileResponseParser.ts]: I wouldn't say well, but it does work afaik.

Maybe write an issue laying out a design for what you mean? I don't think it makes sense to invest any more time into auth for the content api right now, It'd be cool to have for later :).

Comment (t83714) [reply magda-typescript-common/src/OpaCompileResponseParser.ts]: 👍

**Comment by AlexGilleran** [comment]
@t83714 I think it's ready for another look now. I ended up deleting the dataset auth stuff out of the authorization API seeing as (I think?) this registry auth stuff should supercede it, if that's OK.

**Comment by t83714** [comment]
> @t83714 I think it's ready for another look now. I ended up deleting the dataset auth stuff out of the authorization API seeing as (I think?) this registry auth stuff should supercede it, if that's OK.

I think it's for generating a list of approvers (on review screen) for sending notifcations~
Yeah~ Let's remove it~ I think we are a new plan for approval process now

**Pull request merged**